api/v1alpha8/types.go

```diff
@@ -301,39 +301,147 @@ type LoadBalancer struct {
 	Tags []string `json:"tags,omitempty"`
 }
 
-// SecurityGroup represents the basic information of the associated
+// SecurityGroupStatus represents the basic information of the associated
 // OpenStack Neutron Security Group.
-type SecurityGroup struct {
-	Name  string              `json:"name"`
-	ID    string              `json:"id"`
-	Rules []SecurityGroupRule `json:"rules"`
+type SecurityGroupStatus struct {
+	// name of the security group
+	// +optional
+	Name string `json:"name"`
+
+	// id of the security group
+	// +optional
+	ID string `json:"id"`
+
+	// list of security group rules
+	// +optional
+	Rules []SecurityGroupRuleStatus `json:"rules"`
 }
 
-// SecurityGroupRule represent the basic information of the associated OpenStack
+// SecurityGroupRuleSpec represent the basic information of the associated OpenStack
 // Security Group Role.
-type SecurityGroupRule struct {
-	Description     string `json:"description"`
-	ID              string `json:"name"`
-	Direction       string `json:"direction"`
-	EtherType       string `json:"etherType"`
-	SecurityGroupID string `json:"securityGroupID"`
-	PortRangeMin    int    `json:"portRangeMin"`
-	PortRangeMax    int    `json:"portRangeMax"`
-	Protocol        string `json:"protocol"`
-	RemoteGroupID   string `json:"remoteGroupID"`
-	RemoteIPPrefix  string `json:"remoteIPPrefix"`
-}
-
-// Equal checks if two SecurityGroupRules are the same.
-func (r SecurityGroupRule) Equal(x SecurityGroupRule) bool {
-	return (r.Direction == x.Direction &&
-		r.Description == x.Description &&
-		r.EtherType == x.EtherType &&
-		r.PortRangeMin == x.PortRangeMin &&
-		r.PortRangeMax == x.PortRangeMax &&
-		r.Protocol == x.Protocol &&
-		r.RemoteGroupID == x.RemoteGroupID &&
-		r.RemoteIPPrefix == x.RemoteIPPrefix)
+// For now this is only used for the allNodesSecurityGroupRules but when we add
+// other security groups, we'll need to add a validation because
+// Remote* fields are mutually exclusive.
+type SecurityGroupRuleSpec struct {
+	// name of the security group rule.
+	// It's used to identify the rule so it can be patched and will not be sent to the OpenStack API.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// description of the security group rule.
+	// +optional
+	Description string `json:"description"`
+
+	// direction in which the security group rule is applied. The only values
+	// allowed are "ingress" or "egress". For a compute instance, an ingress
+	// security group rule is applied to incoming (ingress) traffic for that
+	// instance. An egress rule is applied to traffic leaving the instance.
+	// +kubebuilder:validation:enum=ingress;egress
+	// +optional
+	Direction string `json:"direction"`
+
+	// etherType must be IPv4 or IPv6, and addresses represented in CIDR must match the
+	// ingress or egress rules.
+	// +kubebuilder:validation:enum=IPv4;IPv6
+	// +optional
+	EtherType string `json:"etherType"`
```

Comment:
Bah, this is genuinely optional. I think this must become the dreaded `*string`, with `omitempty`.